src/main/java/io/stargate/sgv2/jsonapi/service/shredding/Shredder.java

```diff
@@ -93,24 +78,52 @@ public WritableShreddedDocument shred(JsonNode doc, UUID txId) {
     validateDocument(documentLimits, docWithId, docJson);
 
     final WritableShreddedDocument.Builder b =
-        WritableShreddedDocument.builder(new DocValueHasher(), docId, txId, docJson, docWithId);
+        WritableShreddedDocument.builder(docId, txId, docJson, docWithId);
 
-    // And now let's traverse the document, _including DocumentId so it will also
-    // be indexed along with other fields.
+    // And finally let's traverse the document to actually "shred" (build index fields)
     traverse(docWithId, b, JsonPath.rootBuilder());
     return b.build();
   }
 
-  private DocumentId generateDocumentId() {
-    return DocumentId.fromUUID(UUID.randomUUID());
+  /**
+   * Method called to ensure that Document has Document Id (generating id if necessary), and that it
+   * is the very first field in the document (reordering as needed). Note that a new document is
+   * created and returned; input document is never modified.
+   *
+   * @param doc Document to use as the base
+   * @return Document that has _id as its first field
+   */
+  private ObjectNode normalizeDocumentId(JsonNode doc) {
+    // First: see if we have Object Id present or not
+    JsonNode idNode = doc.get(DocumentConstants.Fields.DOC_ID);
+    ObjectNode docWithoutId;
+
+    if (idNode == null) { // No id: generate, doc fine to use as source
+      docWithoutId = (ObjectNode) doc;
+      idNode = generateDocumentId();
+    } else { // Has id, use as-is, but create a copy of Doc without id
+      docWithoutId = objectMapper.createObjectNode();
+      docWithoutId.setAll((ObjectNode) doc);
```

Comment:
Code is doing setAll twice in most case, can the docWithoutId outside if condition and add everything back to doc object itself? By that way the reference to document in the Operation class still holds good. I made a fix before to set the document to WritableShreddedDocument that's not needed.

Reply:
Ok, so, I think I can remove one of `setAll()` cases but will not be able to avoid construction of a new `ObjectNode` in all cases unfortunately.

The problem is that due to the way `LinkedHashMap` ordering works, it is not possible (as far as I know) to reorder existing entries (when instance uses insertion-order which is the case with Jackson): new entries will be appended at the end. So it is not possible to move or insert document id as the first field.

But I think I should avoid one of copies since it is fine to just add all fields, including object id, as long as object id already exists in target `ObjectNode`. So there is no need for intermediate `docWithoutId` after all, I think.
I will make changes and ask you to re-review.